src/main/java/org/jboss/msc/service/ServiceContainerImpl.java

```diff
@@ -101,52 +110,26 @@
     private long shutdownInitiated;
 
     private final List<TerminateListener> terminateListeners = new ArrayList<>(1);
-    private final boolean autoShutdown;
 
-    private static final class ShutdownHookHolder {
-        private static final Set<Reference<ServiceContainerImpl, Void>> containers;
-        private static boolean down = false;
-
-        static {
-            containers = new HashSet<>();
-            doPrivileged(new PrivilegedAction<Void>() {
-                public Void run() {
-                    final Thread hook = new Thread(new Runnable() {
-                        public void run() {
-                            // shut down all services in all containers.
-                            final Set<Reference<ServiceContainerImpl, Void>> set = containers;
-                            final LatchListener listener;
-                            synchronized (set) {
-                                down = true;
-                                listener = new LatchListener(set.size());
-                                for (Reference<ServiceContainerImpl, Void> containerRef : set) {
-                                    final ServiceContainerImpl container = containerRef.get();
-                                    if (container == null || !container.isAutoShutdown()) {
-                                        listener.countDown();
-                                        continue;
-                                    }
-                                    container.addTerminateListener(listener);
-                                    container.shutdown();
-                                }
-                                set.clear();
-                            }
-                            // wait for all services to finish.
-                            for (;;) try {
-                                listener.await();
-                                break;
-                            } catch (InterruptedException e) {
-                            }
-                        }
-                    }, "MSC Shutdown Thread");
-                    hook.setDaemon(false);
-                    Runtime.getRuntime().addShutdownHook(hook);
-                    return null;
+    private static final class ShutdownHookThread extends Thread {
```

Comment:
This thread must also define a `run()` method to call `containerRef.get()` and if it is non-`null` call `shutdown` and `awaitTermination` on it.  Without this, the shutdown hook thread doesn't actually do anything.

Reply:
You are right of course! Thanks.